src/data/polynomial/erase_lead.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.degree.basic
+import data.polynomial.degree.trailing_degree
+
+/-!
+# Erase the leading term of a univariate polynomial
+
+## Definition
+
+* `erase_lead f`: the polynomial `f - leading term of f`
+
+`erase_lead` serves as reduction step in an induction, shaving off one monomial from a polynomial.
+The definition is set up so that it does not mention subtraction in the definition,
+and thus works for polynomials over semirings as well as rings.
+-/
+
+noncomputable theory
+open_locale classical
+
+open polynomial finsupp finset
+
+namespace polynomial
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
+/-- `erase_lead f` for a polynomial `f` is the polynomial obtained by
+subtracting from `f` the leading term of `f`. -/
+def erase_lead (f : polynomial R) : polynomial R :=
+finsupp.erase f.nat_degree f
+
+section erase_lead
+
+lemma erase_lead_support (f : polynomial R) :
+  f.erase_lead.support = f.support.erase f.nat_degree :=
+by convert rfl
+
+lemma erase_lead_coeff (i : ℕ) :
+  f.erase_lead.coeff i = if i = f.nat_degree then 0 else f.coeff i :=
+by convert rfl
+
+@[simp] lemma erase_lead_coeff_nat_degree : f.erase_lead.coeff f.nat_degree = 0 :=
+finsupp.erase_same
+
+lemma erase_lead_coeff_of_ne (i : ℕ) (hi : i ≠ f.nat_degree) :
+  f.erase_lead.coeff i = f.coeff i :=
+finsupp.erase_ne hi
+
+@[simp] lemma erase_lead_zero : erase_lead (0 : polynomial R) = 0 :=
+finsupp.erase_zero _
+
+@[simp] lemma erase_lead_add_monomial_nat_degree_leading_coeff (f : polynomial R) :
+  f.erase_lead + monomial f.nat_degree f.leading_coeff = f :=
+begin
+  ext i,
+  simp only [erase_lead_coeff, coeff_monomial, coeff_add, @eq_comm _ _ i],
+  split_ifs with h,
+  { subst i, simp only [leading_coeff, zero_add] },
+  { exact add_zero _ }
+end
+
+@[simp] lemma erase_lead_add_C_mul_X_pow (f : polynomial R) :
+  f.erase_lead + (C f.leading_coeff) * X ^ f.nat_degree = f :=
+by rw [C_mul_X_pow_eq_monomial, erase_lead_add_monomial_nat_degree_leading_coeff]
+
+@[simp] lemma self_sub_monomial_nat_degree_leading_coeff {R : Type*} [ring R] (f : polynomial R) :
+  f - monomial f.nat_degree f.leading_coeff = f.erase_lead :=
+(eq_sub_iff_add_eq.mpr (erase_lead_add_monomial_nat_degree_leading_coeff f)).symm
+
+@[simp] lemma self_sub_C_mul_X_pow {R : Type*} [ring R] (f : polynomial R) :
+  f - (C f.leading_coeff) * X ^ f.nat_degree = f.erase_lead :=
+by rw [C_mul_X_pow_eq_monomial, self_sub_monomial_nat_degree_leading_coeff]
+
+lemma erase_lead_ne_zero (f0 : 2 ≤ f.support.card) : erase_lead f ≠ 0 :=
```

Comment:
Please add `finset.pred_le_card_erase : erase s a - 1 ≤ card (erase s a)`. Then this lemma will be trivial: `2 - 1 ≤ card (erase_lead f).support` implies `(erase_lead f).support ≠ ∅`.